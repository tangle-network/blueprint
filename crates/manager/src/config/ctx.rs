use crate::config::BlueprintManagerConfig;
use crate::error::Result;
#[cfg(feature = "vm-sandbox")]
use crate::rt::hypervisor::net::NetworkManager;
use blueprint_auth::db::RocksDb;
use std::ops::{Deref, DerefMut};
use tokio::sync::Mutex;

#[cfg(feature = "containers")]
pub struct ContainerContext {
    pub kube_client: Option<kube::Client>,
    pub kube_service_port: Mutex<crate::sdk::utils::PortLock>,
    pub local_ip: std::net::IpAddr,
}

#[cfg(feature = "vm-sandbox")]
pub struct VmContext {
    pub network_manager: NetworkManager,
    pub network_interface: String,
}

pub struct BlueprintManagerContext {
    #[cfg(feature = "containers")]
    pub containers: ContainerContext,
    #[cfg(feature = "vm-sandbox")]
    pub vm: VmContext,
    pub(crate) db: Mutex<Option<RocksDb>>,
    config: BlueprintManagerConfig,
}

impl BlueprintManagerContext {
    /// Create a new `BlueprintManagerContext`
    ///
    /// # Errors
    ///
    /// * Unable to create the necessary directories
    /// * With the `vm-sandbox` enabled
    ///   * Unable to determine the default network interface if not specified in `config`
    /// * With the `containers` feature enabled
    ///   * Unable to connect to the Kubernetes cluster
    ///   * Unable to bind to the [`BlueprintManagerConfig::kube_service_port()`]
    ///   * Unable to determine the system's local IPv4 address
    #[allow(clippy::unused_async)]
    pub async fn new(mut config: BlueprintManagerConfig) -> Result<Self> {
        config.paths.data_dir = std::path::absolute(config.data_dir())?;

        config.verify_directories_exist()?;
        #[cfg(feature = "vm-sandbox")]
        let (network_manager, network_interface) = {
            use blueprint_core::info;

            let interface = config.verify_network_interface()?;

            crate::rt::hypervisor::net::nftables::check_net_admin_capability()?;

            let network_candidates = config
                .vm_sandbox_options
                .default_address_pool
                .hosts()
                .filter(|ip| ip.octets()[3] != 0 && ip.octets()[3] != 255)
                .collect();

            if config.vm_sandbox_options.no_vm {
                info!("Skipping VM image check, running in no-vm mode");
            } else {
                info!("Checking for VM images");
                crate::rt::hypervisor::images::download_image_if_needed(config.cache_dir()).await?;
            }

            (NetworkManager::new(network_candidates).await?, interface)
        };

        #[cfg(feature = "containers")]
        let kube_client = match kube::Client::try_default().await {
            Ok(client) => {
                tracing::debug!("Successfully initialized Kubernetes client");
                Some(client)
            }
            Err(e) => {
                tracing::warn!(
                    "Failed to initialize Kubernetes client (container runtime will be unavailable): {}",
                    e
                );
                None
            }
        };

        Ok(Self {
            #[cfg(feature = "containers")]
            containers: ContainerContext {
                kube_client,
                kube_service_port: Mutex::new(crate::sdk::utils::PortLock::lock(
                    config.kube_service_port(),
                )?),
                local_ip: local_ip_address::local_ip()?,
            },
            #[cfg(feature = "vm-sandbox")]
            vm: VmContext {
                network_manager,
                network_interface,
            },
            // Set in `run_blueprint_manager_with_keystore`
            db: Mutex::new(None),
            config,
        })
    }

    pub async fn db(&self) -> Option<RocksDb> {
        self.db.lock().await.clone()
    }

    pub async fn set_db(&self, db: RocksDb) {
        self.db.lock().await.replace(db);
    }

    #[cfg(feature = "containers")]
    pub async fn kube_service_port(&self) -> u16 {
        let mut guard = self.containers.kube_service_port.lock().await;
        guard.unlock()
    }
}

impl Deref for BlueprintManagerContext {
    type Target = BlueprintManagerConfig;

    fn deref(&self) -> &Self::Target {
        &self.config
    }
}

impl DerefMut for BlueprintManagerContext {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.config
    }
}
