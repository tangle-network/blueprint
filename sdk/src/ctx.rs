//! A set of traits and utilities that provide a common interface for interacting with the Gadget SDK.
//!
//! Usually, when you need access to the SDK, you will need to pass the Context to your jobs/functions. In your code, you will create a struct that encapsulates all the things that you would need from outside world from your job.
//! for example, if you need to interact with the network, you will need to have a network client in your struct. If you need to interact with the database storage, you will need to have a db client in your struct. And so on.
//!
//! This module provides a set of traits that you can implement for your struct to make it a context-aware struct by adding new functionalities to it.
//!
//! # Example
//!
//! ```rust,no_run
//! use gadget_sdk::ctx::KeystoreContext;
//! use gadget_sdk::config::StdGadgetConfigurtion;
//!
//! // This your struct that encapsulates all the things you need from outside world.
//! #[derive(Clone, Debug, KeystoreContext)]
//! struct MyContext {
//!   foo: String,
//!   bar: u64,
//!   #[config]
//!   sdk_config: StdGadgetConfigurtion,
//! }
//! // By deriving KeystoreContext, you can now access the keystore client from your struct.
//!
//! #[job(id = 0, params(who), result(_))]
//! async fn my_job(ctx: &MyContext, who: String) -> Result<String, MyError> {
//!   // Access the keystore client from the context.
//!   let keystore = ctx.keystore();
//!  // Do something with the keystore client.
//!  // ...
//!  Ok(format!("Hello, {}!", who))
//! }
//! ```
//!

use crate::keystore::backend::GenericKeyStore;
// derives
pub use gadget_context_derive::*;

/// `KeystoreContext` trait provides access to the generic keystore from the context.
pub trait KeystoreContext<RwLock: lock_api::RawRwLock> {
    /// Get the keystore client from the context.
    fn keystore(&self) -> Result<GenericKeyStore<RwLock>, crate::config::Error>;
}

/// `GossipNetworkContext` trait provides access to the network client from the context.
pub trait GossipNetworkContext {
    /// Get the Goossip client from the context.
    fn gossip_network(&self) -> &crate::network::gossip::GossipHandle;
}
