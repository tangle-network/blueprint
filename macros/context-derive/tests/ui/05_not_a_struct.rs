use gadget_sdk::ctx::KeystoreContext;

#[derive(KeystoreContext)]
enum MyContext {
    Variant1,
    Variant2,
}

fn main() {}
