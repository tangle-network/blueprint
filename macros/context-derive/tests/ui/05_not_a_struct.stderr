error: Context Extensions traits can only be derived for structs
 --> tests/ui/05_not_a_struct.rs:4:6
  |
4 | enum MyContext {
  |      ^^^^^^^^^
